### ISSUE TITLE
fn:element-with-id doesn't return the element with the id passed as argument

### ISSUE THREAD

**Issue by wsdookadr** [BaseXdb/basex#1364]
According to the XQuery 3.0 specification, an implementation should provide the following four
functions which all appear in the spec under section [14.7 Functions on node identifiers](https://www.w3.org/TR/xpath-functions-30/#fns-on-identifiers) (they're also [mentioned in passing](https://www.w3.org/TR/xquery-31/#id-unordered-expressions) in XQuery 3.1 in section 3.13 so I suppose they're also part of XQuery 3.1):
- `fn:generate-id` (implemented by [FnGenerateId.java](https://github.com/BaseXdb/basex/blob/ab3f83734b0bc6525ee8b5931a2c094197963b2b/basex-core/src/main/java/org/basex/query/func/fn/FnGenerateId.java))
- `fn:element-with-id` (implemented by [FnElementWithId.java](https://github.com/BaseXdb/basex/blob/ab3f83734b0bc6525ee8b5931a2c094197963b2b/basex-core/src/main/java/org/basex/query/func/fn/FnElementWithId.java))
- `fn:id` (implemented by [FnId.java](https://github.com/BaseXdb/basex/blob/ab3f83734b0bc6525ee8b5931a2c094197963b2b/basex-core/src/main/java/org/basex/query/func/fn/FnId.java))
- `fn:idref` (implemented by [FnIdref.java](https://github.com/BaseXdb/basex/blob/ab3f83734b0bc6525ee8b5931a2c094197963b2b/basex-core/src/main/java/org/basex/query/func/fn/FnIdref.java))

The first one generates a unique id for an element in a document. The other three, from what I can see, all serve the same purpose, which is to get back the node with a certain id.

BaseX does indeed provide all four functions, the last three have very similar implementations indeed, they all call the ids() method from [Ids.java](https://github.com/BaseXdb/basex/blob/ab3f83734b0bc6525ee8b5931a2c094197963b2b/basex-core/src/main/java/org/basex/query/func/fn/Ids.java).

I wrote the following testcase(which should be [SSCCE](http://sscce.org/) and [MCVE](http://stackoverflow.com/help/mcve) compliant).

It seems that the following testcase outputs no results. I'm not entirely sure if I'm using the retrieval functions properly.

I also have to mention that some related functions are [`db:node-id`](http://docs.basex.org/wiki/Database_Module#db:node-id) and [`db:open-id`](http://docs.basex.org/wiki/Database_Module#db:open-id) which offer similar functionality, except these work at the database-level (as opposed to document-level) and are BaseX specific. I tried `db:node-id` and `db:open-id` and those two work great.

Would anyone like to brainstorm on this? Thank you

```xquery
let $sample :=
document {
<a>
  <b>
    <c>
    </c>
  </b>
  <d>
    <e>
      <f>
        <g></g>
        <h></h>
      </f>
    </e>
  </d>
</a>
}

let $id := fn:generate-id($sample//f[1])
let $el1 := fn:element-with-id($id, $sample)
let $el2 := fn:id($id, $sample)
let $el3 := fn:idref($id, $sample)

return ($el1,$el2,$el3)
```

**Comment by ChristianGruen** [comment]
> which should be [SSCCE](http://sscce.org/) and [MCVE](http://stackoverflow.com/help/mcve) compliant

Thanks for hinting me at the MCVE acronym ;)

The id/idref handling in XQuery has quite a long history: In XML, IDs could be specified via attributes; placement and name of these attributes are defined in schemas (DTD, XML Schema). – As BaseX can be used without schemas, we decided to use fixed attribute names for ID resolutions (see [XMLToken#isId](https://github.com/BaseXdb/basex/blob/ab3f83734b0bc6525ee8b5931a2c094197963b2b/basex-core/src/main/java/org/basex/util/XMLToken.java#L152-L155): An id reference must (to be detected sa such) contain the substring `ref`, an id must contain the substring `id`. The following example will return element `b` as result:

```xquery
let $sample := document {
  <a>
    <b id='id0'/>
  </a>
}
return fn:id('id0', $sample)
```

A good thing: Queries of this type will be accelerated whenever possible by taking advantage of our attribute index structures. A bad thing: We are not 100% compliant with the specification, because, as indicated above, the names of ID and IDREF attributes should be flexible. – Does this help?

**Comment by ChristianGruen** [comment]
I’m closing this, as it’s no bug…

**Comment by wsdookadr** [comment]
`fn:generate-id` generated an id much like yours, starts with the substring `id` followed by a number.
I tried to replace the substring `id` with `idref` and retrieve the node. I wasn't able to.

But in your example, you're explicitly assigning an id. I had an expectation that I could retrieve the node with a generated id, which I wasn't able to.

But, I must be misunderstanding the use of `fn:generate-id` then. It seems that I'd actually have to go over each node, and use `fn:generate-id` and assign each of them an id if I want to retrieve them based on that later on. 

Yes, I too think it's not a bug, much more likely that I don't understand how these functions are supposed to work. I was expecting them to work like `db:node-id` and `db:open-id` but it seems they're different.

**Comment by ChristianGruen** [comment]
> `fn:generate-id` generated an id much like yours, starts with the substring `id` followed by a number.
I tried to replace the substring `id` with `idref` and retrieve the node. I wasn't able to.
>
> But in your example, you're explicitly assigning an id. I had an expectation that I could retrieve the node with a generated id, which I wasn't able to.

Sounds like a perfectly reasonable expectation ;) To be honest, I don’t know why `fn:generate-id` was introduced (I didn’t have a practical use for it so far). I have looked up the [XQFO 3.1 Specification](https://www.w3.org/TR/xpath-functions-31/#func-generate-id), which says that its primary use is to generate hyperlinks, and that there is indeed no inverse to this function.
